uefi/src/proto/network/http.rs

```diff
@@ -0,0 +1,370 @@
+        } else {
+            ptr::null()
+        } as *mut c_void;
+
+        let mut rx_token = HttpToken {
+            status: Status::NOT_READY,
+            message: &mut rx_msg,
+            ..Default::default()
+        };
+
+        let p = self.protocol.as_mut().expect("no http protocol");
+        p.response(&mut rx_token)?;
+
+        loop {
+            if rx_token.status != Status::NOT_READY {
+                break;
+            }
+            p.poll()?;
+        }
+
+        debug!(
+            "http: response: {} / {:?}",
+            rx_token.status, rx_rsp.status_code
+        );
+
+        if rx_token.status != Status::SUCCESS && rx_token.status != Status::HTTP_ERROR {
+            return Err(rx_token.status.into());
+        };
+
+        debug!("http: headers: {}", rx_msg.header_count);
+        let mut headers: Vec<(String, String)> = Vec::new();
+        for i in 0..rx_msg.header_count {
+            let n;
+            let v;
+            unsafe {
+                n = CStr::from_ptr((*rx_msg.header.add(i)).field_name as *const c_char);
+                v = CStr::from_ptr((*rx_msg.header.add(i)).field_value as *const c_char);
+            }
+            headers.push((
+                n.to_str().unwrap().to_lowercase(),
+                String::from(v.to_str().unwrap()),
+            ));
+        }
+
+        debug!("http: body: {}/{}", rx_msg.body_length, body.len());
+
+        let rsp = HttpHelperResponse {
+            status: rx_rsp.status_code,
+            headers,
+            body: body[0..rx_msg.body_length].to_vec(),
+        };
+        Ok(rsp)
+    }
+
+    /// Receive more body data.
+    pub fn response_more(&mut self) -> uefi::Result<Vec<u8>> {
+        let mut body = vec![0; 16 * 1024];
+        let mut rx_msg = HttpMessage {
+            body_length: body.len(),
+            body: body.as_mut_ptr() as *mut c_void,
+            ..Default::default()
+        };
+
+        let mut rx_token = HttpToken {
+            status: Status::NOT_READY,
+            message: &mut rx_msg,
+            ..Default::default()
+        };
+
+        let p = self.protocol.as_mut().expect("no http protocol");
+        p.response(&mut rx_token)?;
+
+        loop {
+            if rx_token.status != Status::NOT_READY {
+                break;
+            }
+            p.poll()?;
+        }
+
+        debug!("http: response: {}", rx_token.status);
+
+        if rx_token.status != Status::SUCCESS {
+            return Err(rx_token.status.into());
+        };
+
+        debug!("http: body: {}/{}", rx_msg.body_length, body.len());
+
+        Ok(body[0..rx_msg.body_length].to_vec())
+    }
+}
+
+impl Drop for HttpHelper {
+    fn drop(&mut self) {
+        debug!("http: drop");
+        self.protocol = None;
+        if self.binding.is_some() {
+            let b = self.binding.as_mut().unwrap();
+            let _ = b.destroy_child(self.child_handle.unwrap());
+        }
+        self.binding = None;
```

Comment:
I don't see the value add of setting `protocol` and `binding?? to `None` on drop. The structs will be dropped anyway.

Do these fields need to be `Option` in the first place?